### PR TITLE
Performance improvements for large scale computations

### PR DESCRIPTION
**Pull request by micheles**
See https://bugs.launchpad.net/oq-engine/+bug/1462783. The tests are green: https://ci.openquake.org/job/zdevel_oq-risklib/691

### CODE REVIEW

File: openquake/commonlib/source.py [hunk 1]
```diff
@@ -553,17 +536,34 @@ def trt_models(self):
...
                     src.trt_model_id = trt_model.id
-                yield src
+                sources.append(src)
+        return sources
+
+    def get_num_sources(self):
+        """
+        :returns: the total number of sources in the model
+        """
+        return sum(1 for src in self.get_sources())
```
Comment (ptormene) [inline openquake/commonlib/source.py]: You could do len(self.get_sources())

File: openquake/commonlib/source.py [hunk 2]
```diff
@@ -553,17 +536,34 @@ def trt_models(self):
...
+
+    def count_ruptures(self, really=False):
+        """
+        If `really` is True, or if .num_ruptures is zero, update
+        the attribute .num_ruptures in each TRT model.
```
Comment (ptormene) [inline openquake/commonlib/source.py]: Please further explain the meaning of 'really'

File: openquake/commonlib/source.py [hunk 3]
```diff
@@ -650,3 +650,120 @@ def _collect_source_model_paths(smlt):
...
+                             src.__class__.__name__, [], f_time, 0)
+    else:  # only split
+        f_time = 0
+    t1 = time.time()
+    out = []
+    for ss in sourceconverter.split_source(src, sourceprocessor.asd):
+        ss.weight = get_weight(ss)
+        out.append(ss)
+    s_time = time.time() - t1
+    return SourceInfo(src.trt_model_id, src.source_id,
```
Comment (ptormene) [inline openquake/commonlib/source.py]: Please rename to filter_time and split_time

File: openquake/commonlib/source.py [hunk 4]
```diff
@@ -650,3 +650,120 @@ def _collect_source_model_paths(smlt):
...
+        sources_by_trt += (ss.reduce(self.agg_source_info)
+                           if slow_sources else AccumDict())
+
+        # store csm.proctimes
+        source_info_dt = numpy.dtype(
+            [('trt_model_id', int),
+             ('source_id', (str, 20)),
+             ('source_class', (str, 20)),
+             ('split_num', int),
+             ('f_time', float),
```
Comment (ptormene) [inline openquake/commonlib/source.py]: same as above

File: openquake/commonlib/source.py [hunk 5]
```diff
@@ -650,3 +650,120 @@ def _collect_source_model_paths(smlt):
...
+             ('s_time', float)])
+        self.outs.sort(key=lambda o: o[4] + o[5], reverse=True)
+        csm.proctimes = numpy.array(self.outs, source_info_dt)
```
Comment (ptormene) [inline openquake/commonlib/source.py]: Please rename to source_info

**Comment by ptormene** [comment]
Besides the minor notes, LGTM

File: openquake/commonlib/source.py (hunk 1 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

File: openquake/commonlib/source.py (hunk 2 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

File: openquake/commonlib/source.py (hunk 3 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

File: openquake/commonlib/source.py (hunk 4 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

File: openquake/commonlib/source.py (hunk 5 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

**Pull request merged**